File: trunk/SweetHome3D/src/com/eteks/sweethome3d/io/ContentDigestManager.java
/*
 * ContentDigestManager.java 23 mai 2014
 *
 * Sweet Home 3D, Copyright (c) 2014 Emmanuel PUYBARET / eTeks <info@eteks.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
package com.eteks.sweethome3d.io;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLEncoder;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

import com.eteks.sweethome3d.model.Content;
import com.eteks.sweethome3d.tools.ResourceURLContent;
import com.eteks.sweethome3d.tools.SimpleURLContent;
import com.eteks.sweethome3d.tools.URLContent;

/**
 * Manager able to store and compute content digest to compare content data faster.  
 * @author Emmanuel Puybaret
 */
public class ContentDigestManager {
  private static final String  DIGEST_ALGORITHM = "SHA-1";
  private static final byte [] INVALID_CONTENT_DIGEST = {};

  private static ContentDigestManager instance;
  
  private Map<Content, byte []>  contentDigestsCache;
  
  private Map<URLContent, URL>   zipUrlsCache;
  private Map<URL, List<String>> zipUrlEntriesCache;

  private ContentDigestManager() {
    this.contentDigestsCache = new WeakHashMap<Content, byte[]>();
    this.zipUrlsCache = new WeakHashMap<URLContent, URL>();
    this.zipUrlEntriesCache = new WeakHashMap<URL, List<String>>();
  }
  
  /**
   * Returns an instance of this singleton. 
   */
  public static ContentDigestManager getInstance() {
    if (instance == null) {
      synchronized (ContentDigestManager.class) {
        if (instance == null) {
          instance = new ContentDigestManager();
        }
      }
    }
    return instance;
  }

  /**
   * Returns <code>true</code> if the contents in parameter contains the same data,
   * comparing their digest. If the digest of the contents was not 
   * {@linkplain #setContentDigest(Content, byte[]) set} directly, it will be 
   * computed on the fly.
   */
  public boolean equals(Content content1, Content content2) {
    byte [] content1Digest = getContentDigest(content1);
    if (content1Digest == INVALID_CONTENT_DIGEST) {
      return false;
    } else {
      return Arrays.equals(content1Digest, getContentDigest(content2));
    }
  }

  /**
   * Returns <code>true</code> if the digest of the given <code>content</code>
   * is equal to <code>digest</code>.
   */
  public boolean isContentDigestEqual(Content content, byte [] digest) {
    byte [] contentDigest = getContentDigest(content);
    if (contentDigest == INVALID_CONTENT_DIGEST) {
      return false;
    } else {
      return Arrays.equals(contentDigest, digest);
    }
  }

  /**
   * Sets the SHA-1 digest of the given <code>content</code>.
   */
  public synchronized void setContentDigest(Content content, byte [] digest) {
    this.contentDigestsCache.put(content, digest);
  }
  
  /**
   * Returns the SHA-1 digest of the given <code>content</code>, computing it 
   * if it wasn't set.
   */
  public synchronized byte [] getContentDigest(Content content) {
    byte [] digest = this.contentDigestsCache.get(content);
    if (digest == null) {
      try {
        if (content instanceof ResourceURLContent) {
          digest = getResourceContentDigest((ResourceURLContent)content);
        } else if (content instanceof URLContent
                   && !(content instanceof SimpleURLContent)
                   && ((URLContent)content).isJAREntry()) {
          URLContent urlContent = (URLContent)content;
          // If content comes from a home stream
          if (urlContent instanceof HomeURLContent) {
            digest = getHomeContentDigest((HomeURLContent)urlContent);            
          } else {
            digest = getZipContentDigest(urlContent);
          }
        } else {
          digest = computeContentDigest(content);
        }
      } catch (NoSuchAlgorithmException ex) {
        throw new InternalError("No SHA-1 message digest is available");
      } catch (IOException ex) {
        digest = INVALID_CONTENT_DIGEST;
      }
      this.contentDigestsCache.put(content, digest);
    }
    return digest;
  }

  /**
   * Returns the digest of a content coming from a resource file.
   */
  private byte [] getResourceContentDigest(ResourceURLContent urlContent) throws IOException, NoSuchAlgorithmException {
    if (urlContent.isMultiPartResource()) {
      if (urlContent.isJAREntry()) {
        URL zipUrl = urlContent.getJAREntryURL();
        String entryName = urlContent.getJAREntryName();
        int lastSlashIndex = entryName.lastIndexOf('/');
        if (lastSlashIndex != -1) {
          // Consider content is a multi part resource only if it's in a subdirectory
          MessageDigest messageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM);
          String entryDirectory = entryName.substring(0, lastSlashIndex + 1);
          for (String zipEntryName : getZipURLEntries(urlContent)) {
            if (zipEntryName.startsWith(entryDirectory) 
                && !zipEntryName.equals(entryDirectory)
                && isSignificant(zipEntryName)) {
              Content siblingContent = new URLContent(new URL("jar:" + zipUrl + "!/" 
                  + URLEncoder.encode(zipEntryName, "UTF-8").replace("+", "%20")));
              updateMessageDigest(messageDigest, siblingContent);
            }
          }
          return messageDigest.digest();
        } else {
          // Consider the content as not a multipart resource
          return computeContentDigest(urlContent);
        }
      } else {
        // This should be the case only when resource isn't in a JAR file during development
        try {
          MessageDigest messageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM);
          File contentFile = new File(urlContent.getURL().toURI());
          File parentFile = new File(contentFile.getParent());
          File [] siblingFiles = parentFile.listFiles();
          // Sort files to ensure content files are always listed in the same order
          Arrays.sort(siblingFiles);
          for (File siblingFile : siblingFiles) {
            if (!siblingFile.isDirectory()) {
              updateMessageDigest(messageDigest, new URLContent(siblingFile.toURI().toURL()));
            }
          }
          return messageDigest.digest();
        } catch (URISyntaxException ex) {
          IOException ex2 = new IOException();
          ex2.initCause(ex);
          throw ex2;
        }
      }
    } else {
      return computeContentDigest(urlContent);
    }
  }

  /**
   * Returns the digest of a content coming from a home file.
   */
  private byte [] getHomeContentDigest(HomeURLContent urlContent) throws IOException, NoSuchAlgorithmException {
    String entryName = urlContent.getJAREntryName();
    int slashIndex = entryName.indexOf('/');
    // If content comes from a directory of a home file
    if (slashIndex > 0) {
      URL zipUrl = urlContent.getJAREntryURL();
      String entryDirectory = entryName.substring(0, slashIndex + 1);
      MessageDigest messageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM);
      for (String zipEntryName : getZipURLEntries(urlContent)) {
        if (zipEntryName.startsWith(entryDirectory) 
            && !zipEntryName.equals(entryDirectory)
            && isSignificant(zipEntryName)) {
          Content siblingContent = new URLContent(new URL("jar:" + zipUrl + "!/" 
              + URLEncoder.encode(zipEntryName, "UTF-8").replace("+", "%20")));
          updateMessageDigest(messageDigest, siblingContent);    
        }
      }
      return messageDigest.digest();
    } else {
      return computeContentDigest(urlContent);
    }
  }

  /**
   * Returns the digest of the given zip content.
   */
  private byte [] getZipContentDigest(URLContent urlContent) throws IOException, NoSuchAlgorithmException {
    MessageDigest messageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM);
    // Open zipped stream that contains urlContent
    for (String zipEntryName : ContentDigestManager.getInstance().getZipURLEntries(urlContent)) {
      if (isSignificant(zipEntryName)) {
        Content siblingContent = new URLContent(new URL("jar:" + urlContent.getJAREntryURL() + "!/" 
            + URLEncoder.encode(zipEntryName, "UTF-8").replace("+", "%20")));
        updateMessageDigest(messageDigest, siblingContent);
      }
    }
    return messageDigest.digest();
  }
  
  /**
   * Returns <code>true</code> if entry name is significant to distinguish 
   * the data of a content from an other one.
   */
  private boolean isSignificant(String entryName) {
    // Ignore LICENSE.TXT files
    String entryNameUpperCase = entryName.toUpperCase();
    return !entryNameUpperCase.equals("LICENSE.TXT") 
          && !entryNameUpperCase.endsWith("/LICENSE.TXT");
  }

  /**
   * Returns the list of entries contained in <code>zipUrl</code>.
   */
  synchronized List<String> getZipURLEntries(URLContent urlContent) throws IOException {
    URL zipUrl = this.zipUrlsCache.get(urlContent);
    if (zipUrl != null) {
      return this.zipUrlEntriesCache.get(zipUrl); 
    } else {
      zipUrl = urlContent.getJAREntryURL(); 
      for (Map.Entry<URL, List<String>> entry : this.zipUrlEntriesCache.entrySet()) {
        if (zipUrl.equals(entry.getKey())) {
          this.zipUrlsCache.put(urlContent, entry.getKey());
          return entry.getValue();
        }
      }
      List<String> zipUrlEntries = new ArrayList<String>();
      ZipInputStream zipIn = null;
      try {
        // Search all entries of zip url
        zipIn = new ZipInputStream(zipUrl.openStream());
        for (ZipEntry entry; (entry = zipIn.getNextEntry()) != null; ) {
          zipUrlEntries.add(entry.getName());
        }
        // Sort entries to ensure the files of multi part content are always listed 
        // in the same order whatever its source
        Collections.sort(zipUrlEntries);
        // Store retrieved entries in the map with a URL key  
        this.zipUrlEntriesCache.put(zipUrl, zipUrlEntries);
        // Store URL in a map with keys that will be referenced as long as they are needed in the program
        // This second map allows to use a weak hash map for zipUrlEntriesCache that will be cleaned
        // only once all the URLContent objects sharing a same URL are not used anymore 
        this.zipUrlsCache.put(urlContent, zipUrl);
        return zipUrlEntries;
      } finally {
        if (zipIn != null) {
          zipIn.close();
        }
      }
    }
  }

  /**
   * Returns the digest of the given <code>content</code>.
   */
  private byte [] computeContentDigest(Content content) throws IOException, NoSuchAlgorithmException {
    MessageDigest messageDigest = MessageDigest.getInstance(DIGEST_ALGORITHM);
    updateMessageDigest(messageDigest, content);
    return messageDigest.digest();
  }

  /**
   * Updates message digest with the data of the given <code>content</code>.
   */
  private void updateMessageDigest(MessageDigest messageDigest, Content content) throws IOException {
    InputStream in = null;
    try {
      in = content.openStream();
      byte [] buffer = new byte [8192];
      int size; 
      while ((size = in.read(buffer)) != -1) {
        messageDigest.update(buffer, 0, size);
      }
    } finally {
      if (in != null) {
        in.close();
      }
    }
  }
}

File: trunk/SweetHome3DJS/src/ContentDigestManager.js
/*
 * ContentDigestManager.js
 *
 * Sweet Home 3D, Copyright (c) 2023 Emmanuel PUYBARET / eTeks <info@eteks.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// Requires jszip.min.js
//          URLContent.js

/**
 * Manager able to store and compute content digest to compare content data faster.
 * @constructor
 * @private
 * @author Emmanuel Puybaret
 */
function ContentDigestManager() {
  this.contentDigestsCache = {};
  this.permanentContents = {};
}

/**
 * Singleton
 * @private
 */ 
ModelManager.instance = null;

/**
 * Returns an instance of this singleton.
 * @return {ModelManager} 
 */
ContentDigestManager.getInstance = function() {
  if (ContentDigestManager.instance == null) {
    ContentDigestManager.instance = new ContentDigestManager();
  }
  return ContentDigestManager.instance;
}

/**
 * Returns <code>true</code> if the contents in parameter contains the same data,
 * comparing their digest. If the digest of the contents was not
 * {@linkplain #setContentDigest(Content, byte[]) set} directly or in cache,
 * it will return <code>false</code>.
 * @param {URLContent} content1
 * @param {URLContent} content2
 * @return {boolean}
 */
ContentDigestManager.prototype.equals = function(content1, content2) {
    var content1Digest = this.contentDigestsCache [content1.getURL()];
    if (content1Digest == null) {
      return false;
    } else {
      return content1Digest === this.contentDigestsCache [content2.getURL()];
    }
  }

/**
 * Sets the SHA-1 digest of the given <code>content</code>. 
 * It should be used for permanent contents like the ones coming for preferences. 
 * @param {URLContent} content
 * @param {string} digest
 */
ContentDigestManager.prototype.setContentDigest = function(content, digest) {
  this.contentDigestsCache [content.getURL()] = digest;
  this.permanentContents [digest] = content;
}

/**
 * Returns the permanent content which matches the given <code>content</code> 
 * or <code>null</code> if doesn't exist.
 * @param {URLContent} content  a content with its digest already computed 
 * @return {URLContent}
 */
ContentDigestManager.prototype.getPermanentContentDigest = function(content) {
  var contentDigest = this.contentDigestsCache [content.getURL()];
  if (contentDigest !== undefined) {
    var permanentContent = this.permanentContents [contentDigest];
    if (permanentContent !== undefined) {
      return permanentContent;
    }
  } 
  
  return null;
}

/**
 * Returns asynchronously the SHA-1 digest of the given content.
 * @param {URLContent} content
 * @param {digestReady: function, digestError: function} digestObserver
 * @ignored
 */
ContentDigestManager.prototype.getContentDigest = function(content, digestObserver) {
  var contentDigest = this.contentDigestsCache [content.getURL()];
  if (contentDigest === undefined) {
    if (content.isJAREntry()) {
      this.getZipContentDigest(content, digestObserver);
    } else if (content instanceof LocalURLContent) {
      var manager = this;
      content.getBlob({
          blobReady: function(blob) {
            if (blob.type === "application/zip") {
              manager.getZipContentDigest(content, digestObserver);
            } else {
              manager.getURLContentDigest(content, digestObserver);
            }
          },
          blobError: function(status, error) {
            digestObserver.digestError(status, error);
          }    
        });
    } else {
      this.getURLContentDigest(content, digestObserver);
    }
  } else {
    digestObserver.digestReady(content, contentDigest);
  }
}

/**
 * Returns asynchronously the SHA-1 digest of the given content.
 * @param {URLContent} content content containing zipped data
 * @param {digestReady: function, digestError: function} digestObserver
 * @private
 */
ContentDigestManager.prototype.getZipContentDigest = function(content, digestObserver) {
  var manager = this;
  ZIPTools.getZIP(content.isJAREntry() ? content.getJAREntryURL() : content.getURL(), false, {
      zipReady : function(zip) {
        try {
          var entryName = content.isJAREntry() ? content.getJAREntryName() : "";
          var slashIndex = content instanceof HomeURLContent
              ? entryName.indexOf('/')
              : -1;
          var entryDirectory = entryName.substring(0, slashIndex + 1);
          var contentData = new Uint8Array(0);
          var entries = slashIndex > 0 || !(content instanceof HomeURLContent) 
              ? zip.file(new RegExp("^" + entryDirectory + ".*")).sort(function(entry1, entry2) { return entry1.name === entry2.name ? 0 : (entry1.name < entry2.name ? 1 : -1); }) // Reverse order
              : [zip.file(entryName)];
          
          for (var i = entries.length - 1; i >= 0 ; i--) {
            var zipEntry = entries [i];
            if (zipEntry.name !== entryDirectory
                && manager.isSignificant(zipEntry.name)) {
              // Append entry data to contentData
              var entryData = zipEntry.asUint8Array();
              var data = new Uint8Array(contentData.length + entryData.length);
              data.set(contentData);
              data.set(entryData, contentData.length);
              contentData = data;
            }
          }   
          
          manager.computeContentDigest(contentData, function(digest) {
              manager.contentDigestsCache [content.getURL()] = digest;
              digestObserver.digestReady(content, digest);
            });              
        } catch (ex) {
          this.zipError(ex);
        }
      },
      zipError : function(error) {
        digestObserver.digestError(error, error.message);
      }
    });
}

/**
 * Returns asynchronously the SHA-1 digest of the given content.
 * @param {URLContent} content content containing no zipped data 
 * @param {digestReady: function, digestError: function} digestObserver
 * @private
 */
ContentDigestManager.prototype.getURLContentDigest = function(content, digestObserver) {
  var manager = this;
  content.getStreamURL({
      urlReady: function(url) {
        var request = new XMLHttpRequest();
        request.open("GET", url, true);
        request.responseType = "arraybuffer";
        request.addEventListener("load", function() {
            manager.computeContentDigest(request.response, function(digest) {
                manager.contentDigestsCache [content.getURL()] = digest;
                digestObserver.digestReady(content, digest);
              });
          });
        request.send();
      },
      urlError: function(status, error) {
        digestObserver.digestError(status, error);
      }    
    });
}

/**
 * Computes the digest of the given data and calls <code>observer</code> when digest is ready. 
 * @param {Uint8Array} contentData
 * @param {function} observer callback which will receive in parameter the SHA-1 digest of contentData in Base64
 * @private
 */
ContentDigestManager.prototype.computeContentDigest = function(contentData, observer) {
  var crypto = window.msCrypto !== undefined ? window.msCrypto : window.crypto;
  var digest;
  try {
    digest = crypto.subtle.digest("SHA-1", contentData);
  } catch (ex) {
    // Use SHA-256 instead even if secured hash is not needed here
    digest = crypto.subtle.digest("SHA-256", contentData);
  }
  if (digest.then !== undefined) { 
    digest.then(function(hash) {
        observer(btoa(String.fromCharCode.apply(null, new Uint8Array(hash))));
      });
  } else { 
    // IE 11 digest.result is available without promise support but only after a call to setTimeout 
    setTimeout(function() {
        observer(btoa(String.fromCharCode.apply(null, new Uint8Array(digest.result))));      
      });
  }
}

/**
 * Returns <code>true</code> if entry name is significant to distinguish
 * the data of a content from an other one. 
 * @param {string} entryName
 * @return {boolean} 
 * @private
 */
ContentDigestManager.prototype.isSignificant = function(entryName) {
  // Ignore LICENSE.TXT files
  var entryNameUpperCase = entryName.toUpperCase();
  return entryNameUpperCase !== "LICENSE.TXT"
        && entryNameUpperCase.indexOf("/LICENSE.TXT", entryNameUpperCase.length - "/LICENSE.TXT".length) === -1;
}
